Add sample deployment/service account manifest for CAPI

benmoss opened the pull request:
Based on https://notes.elmiko.dev/2020/05/22/kubernetes-autoscaler-capd.html

@detiber @elmiko

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/deployment.yaml:
```diff
@@ -0,0 +1,158 @@
+kind: Deployment
+metadata:
+  name: cluster-autoscaler
+  namespace: kube-system
```
This should probably not be kube-system by default. It might be good to use a variable for this and expect users to run envsubst or similar against this prior to deploying.

elmiko commented on cluster-autoscaler/cloudprovider/clusterapi/deployment.yaml:
```diff
@@ -0,0 +1,158 @@
+        - --cloud-provider=clusterapi
+        - --namespace=default
+        - --scale-down-enabled=true
+        - --scale-down-delay-after-add=10s
+        - --scale-down-delay-after-delete=10s
+        - --scale-down-delay-after-failure=10s
+        - --scale-down-unneeded-time=23s
```
we might want to ease these scale down times back a little. i have used these aggressive timings for demonstration purposes but they are not suited for production. imo, we could drop these in favor of the default values (which are in the minutes), or perhaps change just put the defaults here so people can see them.

detiber replied:
I believe we should probably use a variable for these as well, since these will likely need to be tweaked for "real world" type of environments.

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/deployment.yaml:
```diff
@@ -0,0 +1,158 @@
+        - --scale-down-delay-after-failure=10s
+        - --scale-down-unneeded-time=23s
+        - --balance-similar-node-groups
+        - --max-node-provision-time=10m
```
I believe we should probably use a variable for this as well, since this will likely need to be tweaked depending on which provider is being used for Cluster API.

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/deployment.yaml:
```diff
@@ -0,0 +1,158 @@
+        - --balance-similar-node-groups
+        - --max-node-provision-time=10m
+        - --expendable-pods-priority-cutoff=-10
+        - --max-nodes-total=24
```
I believe we should probably use a variable for this as well, since this will likely need to be tweaked depending on the scale of workload clusters a user might be deploying.

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/deployment.yaml:
```diff
@@ -0,0 +1,158 @@
+  name: cluster-autoscaler
+  namespace: kube-system
+---
+kind: ClusterRole
```
Can you break this into two separate ClusterRoles? One for workload cluster resources and one for management cluster resources?

That will make it simpler to break out into separate manifests for each use case once we have support for connecting to separate workload and management clusters.

elmiko replied:
i like this, great idea

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/deployment.yaml:
```diff
@@ -0,0 +1,158 @@
+        app: cluster-autoscaler
+    spec:
+      containers:
+      - image: us.gcr.io/k8s-artifacts-prod/autoscaling/cluster-autoscaler:v1.18.1
```
I'm not quite sure how we want to handle it yet, but we probably want to consider taking into account for overriding either the image repository/name and the version tag (or both), since users may want to override the registry/image name for testing non-released versions and the version tag will need to correspond to the same minor kubernetes release that is running on the managed workload cluster.

elmiko replied:
this makes way too much sense ;)

elmiko commented:
+1 to variable usage where possible, nice suggestion Jason.

benmoss commented:
I removed the default flags, moved the file to under `cloudprovider/clusterapi/examples`, and split the `ClusterRoles` into management cluster permissions and workload cluster permissions. I just guessed about which permissions belong to which cluster though.

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/examples/deployment.yaml:
```diff
@@ -0,0 +1,167 @@
+subjects:
+- kind: ServiceAccount
+  name: cluster-autoscaler
+  namespace: kube-system
```
```suggestion
  namespace: ${AUTOSCALER_NS}
```

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/examples/deployment.yaml:
```diff
@@ -0,0 +1,167 @@
+kind: ServiceAccount
+metadata:
+  name: cluster-autoscaler
+  namespace: kube-system
```
```suggestion
  namespace: ${AUTOSCALER_NS}
```

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/examples/deployment.yaml:
```diff
@@ -0,0 +1,167 @@
+subjects:
+- kind: ServiceAccount
+  name: cluster-autoscaler
+  namespace: kube-system
```
```suggestion
  namespace: ${AUTOSCALER_NS}
```

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/examples/deployment.yaml:
```diff
@@ -0,0 +1,167 @@
+metadata:
+  name: cluster-autoscaler-management
+rules:
+  - apiGroups:
+    - ""
+    resources:
+    - events
+    verbs:
+    - create
+    - patch
```
I believe this should be moved to `cluster-autoscaler-workload`

benmoss replied:
This was the one I was least sure of, but it makes sense

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/examples/deployment.yaml:
```diff
@@ -0,0 +1,167 @@
+    verbs:
+    - create
+    - patch
+  - apiGroups:
+    - ""
+    resources:
+    - configmaps
+    verbs:
+    - create
+    - delete
+    - get
+    - update
```
I believe this should be moved to cluster-autoscaler-workload

detiber commented on cluster-autoscaler/cloudprovider/clusterapi/examples/deployment.yaml:
```diff
@@ -0,0 +1,167 @@
+    - delete
+    - get
+    - update
+  - apiGroups:
+    - coordination.k8s.io
+    resources:
+    - leases
+    verbs:
+    - create
+    - get
+    - update
```
I believe this should be moved to cluster-autoscaler-workload

detiber commented:
/lgtm

elmiko commented:
/area provider/clusterapi
